ui/components/app/signature-request/signature-request.component.js

```diff
@@ -45,9 +45,85 @@ export default class SignatureRequest extends PureComponent {
       hardwareWalletRequiresConnection,
     } = this.props;
     const { address: fromAddress } = fromAccount;
-    const { message, domain = {} } = JSON.parse(data);
+    const { message, domain = {}, types, primaryType } = JSON.parse(data);
     const { metricsEvent } = this.context;
 
+    const mapType = (msgType, definedType) => {
+      // try and map it to a solidity type
+      if (
+        (definedType.type.indexOf('int') >= 0 ||
+          definedType.type.indexOf('fixed') >= 0) &&
+        msgType === 'number'
+      ) {
+        return true;
+      } else if (
+        (definedType.type.indexOf('bytes') >= 0 ||
+          definedType.type === 'address') &&
+        msgType === 'string'
+      ) {
+        return true;
+      }
+
+      return false;
+    };
+
+    const sanitizeMessage = (msg, baseType) => {
+      if (version === 'V4') {
+        const sanitizedMessage = {};
+        const msgKeys = Object.keys(msg);
+        const msgTypes = Object.values(msg).map((value) => typeof value);
+        const baseTypeType = types[baseType];
+        msgKeys.forEach((msgKey, index) => {
+          const valueType = msgTypes[index];
```

Comment:
Please use `Object.entries` to iterate over the keys and values of an object, rather than using `Object.keys` and `Object.values` and tying them together with the index. It's much easier to read that way, and it avoids concerns of iteration order not being standardized for objects.

Reply:
Also, what is the purpose of looking at the `typeof` of the value? The type of the value is not useful, and is entirely distinct from the type defined in the message.

Reply:
Note that this has not yet been addressed. You're still using `Object.keys` and `Object.values` instead of `Object.entries`.

Reply:
I must have misunderstood; there is no longer a secondary call to `Object.values` on the same object. I was mistaking the type definition object for the message I think.

You still could use `Object.entries` here but it would be a minimal improvement to readability at best.